src/ModelContextProtocol.AspNetCore/McpRouteAwarenessExtension.cs

```diff
@@ -0,0 +1,228 @@
+        });
+
+        // Register the route service as a singleton
+        builder.Services.AddSingleton<RouteAwareToolService>();
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Maps MCP endpoints with route-aware capabilities, creating endpoints for both the global route and all discovered tool routes.
+    /// </summary>
+    /// <param name="endpoints">The endpoint route builder to configure.</param>
+    /// <param name="pattern">The route pattern prefix for the global MCP endpoint.</param>
+    /// <returns>An <see cref="IEndpointConventionBuilder"/> for configuring additional endpoint conventions.</returns>
+    /// <remarks>
+    /// <para>
+    /// This method discovers all tool routes from registered <see cref="McpServerTool"/> instances
+    /// and creates MCP endpoints for each unique route. Tools without route attributes are
+    /// available on the global route and all specific routes.
+    /// </para>
+    /// <para>
+    /// For example, with a global pattern of "mcp" and tools having routes "admin" and "readonly":
+    /// <list type="bullet">
+    /// <item><description>/mcp - serves all tools (global route)</description></item>
+    /// <item><description>/mcp/admin - serves admin-specific tools + global tools</description></item>
+    /// <item><description>/mcp/readonly - serves readonly-specific tools + global tools</description></item>
+    /// </list>
+    /// </para>
+    /// </remarks>
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="endpoints"/> is null.</exception>
+    public static IEndpointConventionBuilder MapMcpWithRouting(this IEndpointRouteBuilder endpoints, [StringSyntax("Route")] string pattern = "")
+    {
+        if (endpoints is null)
+            throw new ArgumentNullException(nameof(endpoints));
+
+        var tools = endpoints.ServiceProvider.GetServices<McpServerTool>();
+        var routeService = endpoints.ServiceProvider.GetService<RouteAwareToolService>();
+
+        // Fallback to standard mapping if route service is not available
+        if (routeService is null)
+        {
+            return endpoints.MapMcp(pattern);
+        }
+
+        // Normalize and register the global route
+        var normalizedGlobalRoute = NormalizePath(pattern);
+        routeService.RegisterGlobalRoute(normalizedGlobalRoute);
+
+        // Discover and register all tool-specific routes
+        foreach (var tool in tools)
+        {
+            if (tool.Routes?.Count > 0)
+            {
+                var toolRoutes = tool.Routes
+                    .Select(route => NormalizePath($"{normalizedGlobalRoute}/{route}"))
+                    .ToArray();
+
+                routeService.RegisterOtherRoutes(toolRoutes);
+            }
+        }
+
+        // Map MCP endpoints for all discovered routes
+        foreach (var discoveredRoute in routeService.OtherRoutes)
+        {
+            endpoints.MapMcp(discoveredRoute);
```

Comment:
I don't think we should be adding any routes that are not specified via the `pattern` argument. If we use attributes, it should only be used to filter which tools are available on routes that were already registered explicitly.